### ISSUE TITLE
input autoFocus causes focus to be emitted before ref

### ISSUE THREAD

**Issue by syranide** [facebook/react#7769]
https://jsfiddle.net/nnwd2c34/

**Comment by zymnytskiy** [comment]
What do mean? onFocus is triggered on load?

**Comment by LeZuse** [comment]
Also came here to report this. Prevents a very usual usecase for `focus`/`select`:
```
<input
  autoFocus
  type="text"
  ref="valueInput"
  onFocus={this.onFocus}
/>

// this.onFocus:
this.refs.valueInput.select();
      ^ refs are not available yet
```

**Comment by syranide** [comment]
@LeZuse You can just use `event.target` as a simpler workaround FYI.

**Comment by LeZuse** [comment]
@syranide ha, true! Thanks!

**Comment by gaearon** [comment]
Still happens on 16.

**Comment by oliviertassinari** [comment]
@syranide The `event.target` workaround works great. For everybody else, you might need `event.currentTarget`, [be aware of the difference](https://stackoverflow.com/questions/5921413/difference-between-e-target-and-e-currenttarget), and pick the one that fit your needs.

**Comment by gaearon** [comment]
Not sure what the right fix would be. Tagging as good first issue for somebody to investigate.

**Comment by RichieAHB** [comment]
The issue here seems to be that `Ref` effects are called after `Update` effects below.

https://github.com/facebook/react/blob/13c5e2b53112c1cd77dd52e028a4bb33dc492d1b/packages/react-reconciler/src/ReactFiberScheduler.js#L312-L342

I'm assuming this is to ensure that anything with a `ref` is guaranteed to be mounted, which seems to make sense. In the off-chance that this *isn't* a guarantee that React provides then we could just flip the order of effects processing. Otherwise, it seems autofocussing would need to be broken away from the mounting process to solve this issue, which seems like an odd thing to do for other reasons.

**Comment by moe091** [comment]
Nobody else has picked up this issue? I'd love to give it a try if not

**Comment by RichieAHB** [comment]
I had a look at this but I feel like it would mean a relatively big change and could probably do with a bit of feedback from @gaearon before pushing on.

**Comment by moe091** [comment]
Yeah good point, it looks like this would require changing the order in which different effects are applied to elements when they are created/updated, and changing that would obviously have a lot of side-effects.

**Comment by RichieAHB** [comment]
Exactly, unless I’ve misunderstood something. If you can find another way
I’d be interested to hear!

**Comment by wuweiweiwu** [comment]
@aweary @gaearon seems like a good first deep dive. Can I pick this up?

**Comment by eragon512** [comment]
hey guys, I'd like to take up this issue if its available

**Comment by RichieAHB** [comment]
Another broad idea to fix this (a few years on) is to swap the order of `Ref` and `Update` but delay hooks until the `Ref` effects have run. No idea how feasible this would be.

**Comment by eps1lon** [comment]
Encountered this issue when using `useImperativeHandle`. The issue is that refs can be set with `useImperativeHandle` after `useLayoutEffects` run. This can become an issue if we execute callbacks in `useLayoutEffect` that use the ref set by `useImperativeHandle` e.g.

https://codesandbox.io/s/dank-darkness-podtr
```js
const Child = React.forwardRef(({ callback }, ref) => {
  React.useLayoutEffect(callback);
  React.useImperativeHandle(ref, () => () =>
    console.log("handled me imperatively")
  );

  return null;
});

function Parent() {
  const childRef = React.useRef(null);
  // crash unless order of effects in Child is swapped
  return <Child callback={() => childRef.current()} ref={childRef} />;
}
```

Tried @RichieAHB proposed fix which resulted in 2 test failures which are desired IMO:

```patch
diff --git a/packages/react-dom/src/__tests__/ReactComponent-test.js b/packages/react-dom/src/__tests__/ReactComponent-test.js
index d3c5cbc1f..450f053e5 100644
--- a/packages/react-dom/src/__tests__/ReactComponent-test.js
+++ b/packages/react-dom/src/__tests__/ReactComponent-test.js
@@ -356,10 +356,10 @@ describe('ReactComponent', () => {
       'start mount',
       'inner 1 render',
       'inner 2 render',
-      'inner 1 componentDidMount',
       'ref 1 got instance 1',
-      'inner 2 componentDidMount',
+      'inner 1 componentDidMount',
       'ref 2 got instance 2',
+      'inner 2 componentDidMount',
       'outer componentDidMount',
       'start update',
       // Previous (equivalent) refs get cleared
@@ -368,10 +368,10 @@ describe('ReactComponent', () => {
       'inner 2 render',
       'ref 1 got null',
       'ref 2 got null',
-      'inner 1 componentDidUpdate',
       'ref 1 got instance 1',
-      'inner 2 componentDidUpdate',
+      'inner 1 componentDidUpdate',
       'ref 2 got instance 2',
+      'inner 2 componentDidUpdate',
       'outer componentDidUpdate',
       'start unmount',
       'outer componentWillUnmount',
diff --git a/packages/react-dom/src/__tests__/ReactComponentLifeCycle-test.js b/packages/react-dom/src/__tests__/ReactComponentLifeCycle-test.js
index fd286f1e6..355112e0b 100644
--- a/packages/react-dom/src/__tests__/ReactComponentLifeCycle-test.js
+++ b/packages/react-dom/src/__tests__/ReactComponentLifeCycle-test.js
@@ -1046,12 +1046,12 @@ describe('ReactComponentLifeCycle', () => {
     expect(log).toEqual([
       'will mount',
       'render',
-      'did mount',
       'ref',
+      'did mount',
 
       'render',
-      'did update',
       'ref',
+      'did update',
     ]);
   });
```
Basically: Refs are called before didUpdate/didMount which should make more sense. If the component already __did__ mount and __did__ update then I should already have the updated refs and not was is currently after-didUpdate.

The issue becomes more apparent with `forwardRef`. If I pass forwarded refs to host components then those are set before layout effects run. But if I manually handle them with `useImperativeHandle` and not carefully check that I place this at the top of my render function I get a slightly different behavior.

Is this something the core team would be interested in or should this be re-opened when a breaking change is scheduled?
